### PR TITLE
AOP Logging

### PR DESCRIPTION
**Pull request by asudhakar635**
AOP Logging module added under microservice utility

### CODE REVIEW

File: ms-aop-logging/src/main/java/com/smarsh/microservice/aop/logs/CustomLogger.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2019 Smarsh Inc.
```
Comment (shobhitshrivastava) [inline ms-aop-logging/src/main/java/com/smarsh/microservice/aop/logs/CustomLogger.java]: Is this the correct copyright format? Should the author name also be added?

File: ms-aop-logging/src/main/java/com/smarsh/microservice/aop/logs/CustomLogger.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+   */
+  @Around(
+      "execution(* *(..)) && @annotation(com.smarsh.microservice.aop.logs.LogInfoExecutionTime)")
+  public Object aroundInfo(ProceedingJoinPoint point) throws Throwable {
+    long start = System.currentTimeMillis();
+    Object result = point.proceed();
+
+    LOGGER.info("{} took {} ms to execute and return {}",
+        MethodSignature.class.cast(point.getSignature()).getMethod().getName(),
+        System.currentTimeMillis() - start, result);
```
Comment (shobhitshrivastava) [inline ms-aop-logging/src/main/java/com/smarsh/microservice/aop/logs/CustomLogger.java]: should we log the result value too? I doubt if it is relevant to the logging purposes here.

File: ms-aop-logging/src/main/java/com/smarsh/microservice/aop/logs/CustomLogger.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+   */
+  @Around(
+      "execution(* *(..)) && @annotation(com.smarsh.microservice.aop.logs.LogDebugExecutionTime)")
+  public Object aroundDebug(ProceedingJoinPoint point) throws Throwable {
+    long start = System.currentTimeMillis();
+    Object result = point.proceed();
+
+    LOGGER.debug("{} took {} ms to execute and return {}",
+        MethodSignature.class.cast(point.getSignature()).getMethod().getName(),
+        System.currentTimeMillis() - start, result);
```
Comment (shobhitshrivastava) [inline ms-aop-logging/src/main/java/com/smarsh/microservice/aop/logs/CustomLogger.java]: Same as above.

File: ms-aop-logging/src/main/java/com/smarsh/microservice/aop/logs/CustomLogger.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+   * This method provides annotation for printing values after method return under info logs.
+   *
+   * @param joinPoint Method Reference
+   * @param result Method Return Value
+   */
+  @AfterReturning(value =
+      "execution(* *(..)) && @annotation(com.smarsh.microservice.aop.logs.LogInfoAfterReturn)",
+      returning = "result")
+  public void afterReturningInfo(JoinPoint joinPoint, Object result) {
+    if (result != null && result != "") {
```
Comment (shobhitshrivastava) [inline ms-aop-logging/src/main/java/com/smarsh/microservice/aop/logs/CustomLogger.java]: Why we don't want to print an empty string? This looks like a special case, and may also be a valid result in some cases.

File: ms-aop-logging/src/main/java/com/smarsh/microservice/aop/logs/CustomLogger.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+   * This method provides annotation for printing values after method return under debug logs.
+   *
+   * @param joinPoint Method Reference
+   * @param result Method Return Value
+   */
+  @AfterReturning(value =
+      "execution(* *(..)) && @annotation(com.smarsh.microservice.aop.logs.LogDebugAfterReturn)",
+      returning = "result")
+  public void afterReturningDebug(JoinPoint joinPoint, Object result) {
+    if (result != null && result != "") {
```
Comment (shobhitshrivastava) [inline ms-aop-logging/src/main/java/com/smarsh/microservice/aop/logs/CustomLogger.java]: same as above.

File: ms-aop-logging/src/test/java/com/smarsh/microservice/aop/logs/CustomLoggerTest.java [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+import java.util.List;
+
+/**
+ * @author asudhakar
+ * @since 1.0
+ */
+public class CustomLoggerTest {
+
+  private AspectProxyClass aspectProxy;
+  private AspectProxyClass aspectProxyClass;
```
[the remark this answers is not in the document]
Comment (asudhakar635) [reply ms-aop-logging/src/test/java/com/smarsh/microservice/aop/logs/CustomLoggerTest.java]: Addressed

File: ms-aop-logging/src/test/java/com/smarsh/microservice/aop/logs/CustomLoggerTest.java [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    aspectProxyClass = new AspectProxyClass();
+    AspectJProxyFactory aspectJProxyFactory = new AspectJProxyFactory(aspectProxyClass);
+    CustomLogger customLogger = Mockito.spy(new CustomLogger());
+    aspectJProxyFactory.addAspect(customLogger);
+    aspectProxy = aspectJProxyFactory.getProxy();
+    logger = Mockito.mock(Logger.class);
+    setFinalStatic(CustomLogger.class.getDeclaredField("LOGGER"), logger);
+  }
+
+  static void setFinalStatic(Field field, Object newValue) throws Exception {
```
[the remark this answers is not in the document]
Comment (asudhakar635) [reply ms-aop-logging/src/test/java/com/smarsh/microservice/aop/logs/CustomLoggerTest.java]: Addressed

**Comment by shobhitshrivastava** [comment]
Looks good to me

**Pull request merged**